### ISSUE TITLE
Consider allowing Porter API parameters to be passed via URL query strings

### ISSUE THREAD

**Issue by derekpierre** [nucypher/nucypher#2735]
Related to https://github.com/derekpierre/nucypher/pull/5.

Prior discussion on Discord with @vepkenez. 

```

damonC — Yesterday at 9:40 PM
@derek made a https://github.com/derekpierre/nucypher/pull/5  all this stuff is standard and web developers deal with long URLs all the time

---

derek — Today at 1:01 PM
Thanks for the PR :thumbsup: !

Question: how do I correctly provide a list as a parameter? i.e. for include/exclude ursulas parameters. Perhaps I am providing it incorrectly...?

Trying the URL parameter approach with the fix, I get a 400, and a message that include_ursulas parameter is not a valid list
>  r = requests.get('http://127.0.0.1:9155/get_ursulas?quantity=2&duration_periods=4&include_ursulas=["0xC23F4F463c779Dd27c08939dC495b8b35C952c89"]')
>  r.text
Out[211]: "{'include_ursulas': ['Not a valid list.']}"


and the request body has include_ursulas as a string
>>>>> request body: {'quantity': '2', 'duration_periods': '4', 'include_ursulas': '["0xC23F4F463c779Dd27c08939dC495b8b35C952c89"]'}


Providing multiple values yields a 400 as well:
>  r = requests.get('http://127.0.0.1:9155/get_ursulas?quantity=2&duration_periods=4&include_ursulas=0xC23F4F463c779Dd27c08939dC495b8b35C952c89&include_ursulas=0xbdA6Ba3Ad624582495C55451A')

and yields the following request body with a single string entry value for include_ursulas
>>>>> request body: {'quantity': '2', 'duration_periods': '4', 'include_ursulas': '0xC23F4F463c779Dd27c08939dC495b8b35C952c89'}

---

damonC — Today at 1:04 PM
Hmm it may be slightly complex… https://stackoverflow.com/questions/57914396/list-of-query-params-with-flask-request-args
I didn’t try include_ursulas[]=…
That is a common way to do it… but might need that getlist stuff… which is no fun
```

**Comment by vepkenez** [comment]
Maybe the solution to avoiding `getlist()` is to change the format for list type data to something like 
`0x00000000000,0x000000000000,0x000000000000`
It's gross but simple.

**Comment by derekpierre** [comment]
👍 Yep, that's the format I used when I got lists to work.

And to differentiate single-entry lists from single value strings, the comma would still need to be provided eg. `http://127.0.0.1:9155/get_ursulas?quantity=5&duration_periods=4&include_ursulas=0xB04FcDF9327f65AB0107Ea95b78BB200C07FA752,&exclude_ursulas=0x5cF1703A1c99A4b42Eb056535840e93118177232,"

**Comment by vepkenez** [comment]
Why would the comma be needed?
Can't we handle it in the serializer field as a `return [x for x in input.split(',')]`?

**Comment by derekpierre** [comment]
I'm intrigued but I think (?) it is too late at that point - interested to hear more. The fields need to get the type they expect and `include_ursulas` / `exclude_ursulas` expects a list of strings - see https://github.com/nucypher/nucypher/blob/porter/nucypher/utilities/porter/control/specifications/porter_schema.py#L70

This is what I was thinking - https://github.com/derekpierre/nucypher/commit/5c73e6faa80503533196da2eed25704ac43ebcbf.

**Comment by derekpierre** [comment]
I guess we would have to change the field to just be a string, and then derive the relevant object from the string...? But then that also affects the data possibly being passed...do we want to just get rid of the payload data component then and only allow URL parameters?

**Comment by vepkenez** [comment]
I was thinking maybe we make a Marshmallow field called "StringList" and it always takes its input as a string with some defined separator and outputs a list.

It is highly weird to handle body payload in a GET request and as @piotr-roslaniec points out, considered suspicious/improper by some browsers.

**Comment by vepkenez** [comment]
Actually now that I think of it, we could make a field called StringList that could take input as a string _or_ a list... and output a list.  Imma try and throw this together real quick.

**Comment by derekpierre** [comment]
If we move forward with this, the parameters for the JSONRPC controller endpoint would also be affected since it would also have to act the same i.e. request provides a string of comma-separated ursulas for `include_ursulas`/`exclude_ursulas`. Is this ok?

> Actually now that I think of it, we could make a field called StringList that could take input as a string or a list
I think I would rather keep String and List separate...no use over complicating it.

**Comment by piotr-roslaniec** [comment]
> I think I would rather keep String and StringList separate...no use over complicating it; keeps things compartmentalized smile . Fields returning multiple types can be confusing for the APIs using it.

Maybe we could find a middle ground here: denote API endpoint with a single return type `StringList`, but implement `StringList` in a way that accepts list with a single element and no separator - Array of one element is still an array. This way we could achieve robustness without compromising on readability or correctness.

So we would accept all of the below:
```
0xmystring
0xmystring,
0xmystring,...,0xmystring
```
This conforms to the "[robustness principle](https://www.wikiwand.com/en/Robustness_principle)":
"(...) programs that send messages to other machines (...) should conform completely to the specifications, but programs that receive messages should accept non-conformant input as long as the meaning is clear."

**Comment by derekpierre** [comment]
@piotr-roslaniec see PR #2746 .

**Comment by derekpierre** [comment]
Closed via #2746 .
